packages/framework/src/Model/Watchdog/WatchdogRepository.php

```diff
@@ -0,0 +1,155 @@
+
+        if ($watchdog === null) {
+            throw new WatchdogNotFoundException($id);
+        }
+
+        return $watchdog;
+    }
+
+    /**
+     * @return \Doctrine\ORM\QueryBuilder
+     */
+    protected function getQueryBuilder(): QueryBuilder
+    {
+        return $this->em->createQueryBuilder()
+            ->select('w')
+            ->from(Watchdog::class, 'w');
+    }
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Product\Product $product
+     * @param string $email
+     * @param int $domainId
+     * @return \Shopsys\FrameworkBundle\Model\Watchdog\Watchdog|null
+     */
+    public function findByProductUuidEmailAndDomainId(Product $product, string $email, int $domainId): ?Watchdog
+    {
+        return $this->getQueryBuilder()
+            ->where('w.product = :product')
+            ->andWhere('w.email = :email')
+            ->andWhere('w.domainId = :domainId')
+            ->setParameter('product', $product)
+            ->setParameter('email', $email)
+            ->setParameter('domainId', $domainId)
+            ->getQuery()->getOneOrNullResult();
+    }
+
+    /**
+     * @param string $locale
+     * @return \Doctrine\ORM\QueryBuilder
+     */
+    public function getWatchdogProductsQueryBuilder(string $locale): QueryBuilder
+    {
+        return $this->getWatchdogRepository()
+            ->createQueryBuilder('w')
+            ->select('IDENTITY(w.product) as productId')
+            ->addSelect('pt.name as productName')
+            ->addSelect('p.catnum as productCatnum')
+            ->addSelect('COUNT(w.product) as watchdogCount')
+            ->leftJoin('w.product', 'p')
+            ->leftJoin('p.translations', 'pt', Join::WITH, 'pt.locale = :locale')
+            ->setParameter('locale', $locale)
+            ->groupBy('w.product, pt.name, p.catnum')
+            ->orderBy('MAX(w.createdAt)', 'DESC');
+    }
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Product\Product $product
+     * @return \Doctrine\ORM\QueryBuilder
+     */
+    public function getWatchdogsByProductQueryBuilder(Product $product): QueryBuilder
+    {
+        return $this->getWatchdogRepository()
+            ->createQueryBuilder('w')
+            ->where('w.product = :product')
+            ->setParameter('product', $product);
+    }
+
+    /**
+     * @return \Shopsys\FrameworkBundle\Model\Watchdog\Watchdog|null
+     */
+    public function findNextWatchdogToSend(): ?Watchdog
+    {
+        $queryBuilder = $this->getWatchdogRepository()
+            ->createQueryBuilder('w')
+            ->leftJoin('w.product', 'p')
+            ->leftJoin(ProductStock::class, 'ps', Join::WITH, 'ps.product = p')
+            ->where('p.sellingDenied = FALSE')
+            ->andWhere('p.calculatedSellingDenied = FALSE')
+            ->andWhere('p.hidden = FALSE')
+            ->andWhere('p.productType != :productType')
+            ->setParameter('productType', ProductTypeEnum::TYPE_INQUIRY)
```

Comment:
What about hidden on domain and sale exclusion on domain?
In my opinion, this approach is error-prone when some changes are made in the logic of how a product is considered sellable.

I would consider somehow use the logic already used elsewhere - in cart should be probably some check if the product can be bought